core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,182 @@
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+#define KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Atomic.hpp>
+
+#include <functional>
+
+namespace Kokkos {
+namespace Experimental {
+
+template <typename T>
+class MaybeReferenceCountedPtr {
+ protected:
+  MaybeReferenceCountedPtr(T* value_ptr)
+      : m_value_ptr(value_ptr), m_control(nullptr) {}
+
+  template <class Deleter>
+  MaybeReferenceCountedPtr(T* value_ptr, Deleter deleter)
+      : m_value_ptr(value_ptr), m_control(new Control{std::move(deleter), 1}) {}
+
+ public:
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      MaybeReferenceCountedPtr&& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+    other.m_value_ptr = nullptr;
+    other.m_control   = nullptr;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      const MaybeReferenceCountedPtr& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      MaybeReferenceCountedPtr&& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr       = other.m_value_ptr;
+      other.m_value_ptr = nullptr;
+      m_control         = other.m_control;
+      other.m_control   = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      const MaybeReferenceCountedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr = other.m_value_ptr;
+      m_control   = other.m_control;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (is_reference_counted())
+        Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~MaybeReferenceCountedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_value_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept {
+    KOKKOS_EXPECTS(bool(*this));
+    return *get();
+  }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the MaybeReferenceCountedPtr manages the lifetime of the
+  // object
+  KOKKOS_FUNCTION bool is_reference_counted() const noexcept {
+    return m_control != nullptr;
+  }
+
+ private:
+  KOKKOS_FUNCTION void cleanup() noexcept {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    // If m_counter is set, then this instance is responsible for managing the
+    // objects pointed to by m_counter and m_value_ptr.
+    if (is_reference_counted()) {
+      KOKKOS_EXPECTS(m_control->m_deleter);
```

Comment:
I'm not sure what programming error this protects against.  I don't see a way to assign to `m_control` a pointer to a `Control` object w/o a `deleter`.  This is protecting against a class invariant (as opposed to line 112, which protects against a pre-condition).

Reply:
Removed.